File: apps/java/src/test/java/chapter09/StandardizerTest.java
package chapter09;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.within;

import chapter02.Features;
import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

class StandardizerTest {
  private static final List<String> COLUMNS = List.of("RM", "LSTAT");

  private static Features row(double rm, double lstat) {
    return new Features(COLUMNS, new double[] {rm, lstat});
  }

  @Test
  @DisplayName("訓練データから列ごとの平均と標準偏差を求める")
  void fitsMeansAndStds() {
    var standardizer = Standardizer.fit(List.of(row(1, 10), row(2, 10), row(3, 40)));

    assertThat(standardizer.means().get("RM")).isCloseTo(2.0, within(1e-12));
    assertThat(standardizer.means().get("LSTAT")).isCloseTo(20.0, within(1e-12));
    assertThat(standardizer.stds().get("RM")).isCloseTo(Math.sqrt(2.0 / 3), within(1e-12));
    assertThat(standardizer.stds().get("LSTAT")).isCloseTo(Math.sqrt(200.0), within(1e-12));
  }

  @Test
  @DisplayName("訓練データの平均と標準偏差で別のデータを標準化する")
  void transformsOtherData() {
    var standardizer = new Standardizer(Map.of("RM", 2.0), Map.of("RM", 0.5));
    var other = List.of(1.0, 2.0, 4.0).stream().map(Samples::rm).toList();

    var standardized = standardizer.transform(other);

    assertThat(standardized).extracting(f -> f.value("RM")).containsExactly(-2.0, 0.0, 4.0);
  }

  @Test
  @DisplayName("すべて同じ値の列は 0 にする")
  void constantColumnBecomesZero() {
    var x = List.of(1.0, 1.0, 1.0).stream().map(Samples::rm).toList();

    var standardized = Standardizer.fit(x).transform(x);

    assertThat(standardized).extracting(f -> f.value("RM")).containsExactly(0.0, 0.0, 0.0);
  }

  @Test
  @DisplayName("標準化する列に無い列はそのまま残す")
  void keepsOtherColumns() {
    var standardizer = new Standardizer(Map.of("RM", 2.0), Map.of("RM", 0.5));

    var standardized = standardizer.transform(row(3, 7));

    assertThat(standardized).isEqualTo(row(2, 7));
  }
}

File: apps/fsharp/tests/MachineLearning.Tests/Chapter09/StandardizerTest.fs
module MachineLearning.Tests.Chapter09.StandardizerTest

open Xunit
open MachineLearning.Chapter09.Standardizer

let rows (column: string) (values: float list) : Map<string, float> list =
    values |> List.map (fun value -> Map.ofList [ column, value ])

[<Fact>]
let ``訓練データから列ごとの平均と標準偏差を求める`` () =
    let train =
        [
            Map.ofList [ "RM", 1.0; "LSTAT", 10.0 ]
            Map.ofList [ "RM", 2.0; "LSTAT", 10.0 ]
            Map.ofList [ "RM", 3.0; "LSTAT", 40.0 ]
        ]

    let standardizer = fitStandardizer [ "RM"; "LSTAT" ] train

    Assert.Equal<Map<string, float>>(Map.ofList [ "RM", 2.0; "LSTAT", 20.0 ], standardizer.Means)
    Assert.Equal(sqrt (2.0 / 3.0), standardizer.Stds["RM"], 12)
    Assert.Equal(sqrt 200.0, standardizer.Stds["LSTAT"], 12)

[<Fact>]
let ``訓練データの平均と標準偏差で、指定した列だけを標準化する`` () =
    let train = rows "RM" [ 1.0; 2.0; 3.0 ]
    let standardizer = fitStandardizer [ "RM" ] train

    let test =
        [ Map.ofList [ "RM", 4.0; "LSTAT", 7.0 ] ] |> transformStandardized standardizer

    Assert.Equal(2.0 / sqrt (2.0 / 3.0), test.Head["RM"], 12)
    Assert.Equal(7.0, test.Head["LSTAT"])

[<Fact>]
let ``標準偏差が 0 の列は 0 にする`` () =
    let train = rows "CHAS" [ 1.0; 1.0 ]

    let standardized = train |> transformStandardized (fitStandardizer [ "CHAS" ] train)

    Assert.Equal<float list>([ 0.0; 0.0 ], standardized |> List.map (fun row -> row["CHAS"]))
